### PR TITLE
WIP Add change_owner and change_permissions to Client.

### PR DESCRIPTION
**Pull request by ihodes**
This is a WIP (still need to add tests). 

Is this a reasonable approach + appropriate naming? Is there anything else you'd like changed?

Thanks!

### CODE REVIEW

File: hdfs/client.py [hunk 1]
```diff
@@ -602,6 +602,31 @@ def rename(self, hdfs_src_path, hdfs_dst_path):
...
+
+    :param hdfs_path: HDFS path.
+    :param owner: New owner for file.
+    :param group: Optional, new group for file.
+    """
+    if group is None:
+      res = self._set_owner(hdfs_path, owner=owner)
+    else:
+      res = self._set_owner(hdfs_path, owner=owner, group=group)
+    if not res.json()['boolean']:
```
Comment (mtth) [inline hdfs/client.py]: Could you check that the response contains a JSON object? I believe it contains an empty body.

File: hdfs/client.py [hunk 2]
```diff
@@ -602,6 +602,31 @@ def rename(self, hdfs_src_path, hdfs_dst_path):
...
+    """Change the owner of file.
+
+    :param hdfs_path: HDFS path.
+    :param owner: New owner for file.
+    :param group: Optional, new group for file.
+    """
+    if group is None:
+      res = self._set_owner(hdfs_path, owner=owner)
+    else:
+      res = self._set_owner(hdfs_path, owner=owner, group=group)
```
Comment (mtth) [inline hdfs/client.py]: Could you check that we need to handle these cases separately? (I.e. that passing `group=None` doesn't have the desired effect in the first case as well.)

File: hdfs/client.py [hunk 3]
```diff
@@ -602,6 +602,31 @@ def rename(self, hdfs_src_path, hdfs_dst_path):
         self.resolve(hdfs_src_path), hdfs_dst_path
       )
 
+  def change_owner(self, hdfs_path, owner, group=None):
+    """Change the owner of file.
+
+    :param hdfs_path: HDFS path.
+    :param owner: New owner for file.
+    :param group: Optional, new group for file.
+    """
```
Comment (mtth) [inline hdfs/client.py]: Nitpick: empty line at the end of each docstring (same below).

**Comment by mtth** [comment]
Cool! Thanks for doing this. A few comments inline; names sound good at first glance, will think about it.

**Comment by ihodes** [comment]
Super, thank you for the review, I think I addressed your comments. 

I didn't get my change_owner tests to run due to permissions issues; I was wondering if you might have a preferred way to handle this?

The error being `{"RemoteException":{"message":"Non-super user cannot change owner","exception":"AccessControlException","javaClassName":"org.apache.hadoop.security.AccessControlException"}}`. When you run the tests in your environment, is the test user a superuser? I just ran these on the cluster I have access to… maybe I should set up the test cluster you document? (Or maybe these tests just magically pass on TravisCI!)

File: hdfs/client.py [hunk 4]
```diff
@@ -602,6 +602,30 @@ def rename(self, hdfs_src_path, hdfs_dst_path):
...
 
+  def change_owner(self, hdfs_path, owner, group=None):
+    """Change the owner of file.
+
+    :param hdfs_path: HDFS path.
+    :param owner: New owner for file.
+    :param group: Optional, new group for file.
+
+    """
+    res = self._set_owner(hdfs_path, owner=owner, group=group)
```
Comment (mtth) [inline hdfs/client.py]: I think you can simplify this line and the next to:
```python
if not self._set_owner(hdfs_path, owner=owner, group=group):
   # ...
```

File: test/test_client.py [hunk 5]
```diff
@@ -504,6 +504,52 @@ def test_missing(self):
...
+    self.client.change_owner('foo', new_owner)
+    status = self.client.status('foo')
+    status = self.client.status('foo')
+    eq_(status['owner'], new_owner)
+
+  def test_file(self):
+    new_owner = 'newowner'
+    self.client.write('foo', 'hello, world!')
+    status = self.client.status('foo')
+    print status
```
Comment (mtth) [inline test/test_client.py]: You forgot to remove this?

File: test/test_client.py [hunk 6]
```diff
@@ -504,6 +504,52 @@ def test_missing(self):
...
 
 
+class TestChangeOwner(_TestSession):
+
+  def test_directory(self):
+    new_owner = 'newowner'
+    self.client._mkdirs('foo')
+    self.client.change_owner('foo', new_owner)
+    status = self.client.status('foo')
+    status = self.client.status('foo')
```
Comment (mtth) [inline test/test_client.py]: Why is this line duplicated? (Same below.)

File: test/test_client.py [hunk 7]
```diff
@@ -504,6 +504,52 @@ def test_missing(self):
...
+    eq_(status['owner'], new_owner)
+
+  def test_file(self):
+    new_owner = 'newowner'
+    self.client.write('foo', 'hello, world!')
+    status = self.client.status('foo')
```
Comment (mtth) [inline test/test_client.py]: You don't seem to use this `status` call.

File: test/test_client.py [hunk 8]
```diff
@@ -504,6 +504,52 @@ def test_missing(self):
...
 
+class TestChangeOwner(_TestSession):
+
+  def test_directory(self):
+    new_owner = 'newowner'
+    self.client._mkdirs('foo')
+    self.client.change_owner('foo', new_owner)
+    status = self.client.status('foo')
+    status = self.client.status('foo')
+    eq_(status['owner'], new_owner)
```
Comment (mtth) [inline test/test_client.py]: In the case (though unlikely) where the original owner is `'newowner'`, this test won't do anything. Could you add a change to a second (different) user to cover all cases? (Same for all other tests.)

**Comment by mtth** [comment]
Great job on setting up the tests! Looks better, a couple general comments:
+ Could you add logging to each function (similarly to how it's done in others)?
+ Could you rename them to `set_owner` and `set_permissions`? (After thinking about it a bit, I feel these'll be more straightforward.)

(A few more inline.)

**Comment by ihodes** [comment]
Thanks for the feedback (and sorry for the weird debugging left in the tests)!

**Comment by ihodes** [comment]
It looks like the tests pass, but the python2.6 version of them timed out on the Hadoop download. Would you mind restarting the build on Travis?

**Comment by mtth** [comment]
Merged, thanks a lot!

**Comment by ihodes** [comment]
Thank you! When you get a chance, no rush, would you mind cutting a release? It'd be awesome to be able to use this downstream in Ibis.

**Comment by mtth** [comment]
@ihodes - https://pypi.python.org/pypi/hdfs/1.3.0 :).

**Comment by ihodes** [comment]
Hah, way ahead of me, thank you :)

**Comment by ihodes** [comment]
Way ahead of me, sorry! Thanks :)

**Pull request merged**